src/core/qgsproject.cpp

```diff
@@ -1141,20 +1115,15 @@ void QgsProject::cleanTransactionGroups( bool force )
 bool QgsProject::readLayer( const QDomNode& layerNode )
 {
   QList<QDomNode> brokenNodes;
-  QList< QPair< QgsVectorLayer*, QDomElement > > vectorLayerList;
-  if ( addLayer( layerNode.toElement(), brokenNodes, vectorLayerList ) )
+  if ( addLayer( layerNode.toElement(), brokenNodes ) )
   {
     // have to try to update joins for all layers now - a previously added layer may be dependent on this newly
     // added layer for joins
     QVector<QgsVectorLayer*> vectorLayers = layers<QgsVectorLayer*>();
     Q_FOREACH ( QgsVectorLayer* layer, vectorLayers )
     {
-      processLayerJoins( layer );
-    }
-
-    if ( !vectorLayerList.isEmpty() )
-    {
-      emit readMapLayer( vectorLayerList.at( 0 ).first, vectorLayerList.at( 0 ).second );
+      // TODO: should be only done later - and with all layers (other layers may have referenced this layer)
```

Comment:
Kind of... the logic simply is not entirely correct, but I did not want to go further and rework also handling of bad layers at the same time :-)

I would like to change how bad layers are handled at some point - instead of requiring that user fixes all the layers, I want to allow invalid layers in the project. Internally in QgsProject we can still keep the layer's XML definition so the layer config is not lost even if we do not fix it immediately. Invalid layers could be shown in the layer tree (e.g. with a warning icon and/or red text color) and it would be up to user to fix the issue in GUI (or leave it as is).